gpiozero/pins/mock.py

```diff
@@ -275,3 +275,141 @@ def _set_frequency(self, value):
         if value is None:
             self._change_state(0.0)
 
+
+class MockSPIClockPin(MockPin):
+    """
+    This derivative of :class:`MockPin` is intended to be used as the clock pin
+    of a mock SPI device. It is not intended for direct construction in tests;
+    rather, construct a :class:`MockSPIDevice` with various pin numbers, and
+    this class will be used for the clock pin.
+    """
+    def __init__(self, number):
+        super(MockSPIClockPin, self).__init__()
+        if not hasattr(self, 'spi_devices'):
+            self.spi_devices = []
+
+    def _set_state(self, value):
+        super(MockSPIClockPin, self)._set_state(value)
+        for dev in self.spi_devices:
+            dev.on_clock()
+
+
+class MockSPISelectPin(MockPin):
+    """
+    This derivative of :class:`MockPin` is intended to be used as the select
+    pin of a mock SPI device. It is not intended for direct construction in
+    tests; rather, construct a :class:`MockSPIDevice` with various pin numbers,
+    and this class will be used for the select pin.
+    """
+    def __init__(self, number):
+        super(MockSPISelectPin, self).__init__()
+        if not hasattr(self, 'spi_device'):
+            self.spi_device = None
+
+    def _set_state(self, value):
+        super(MockSPISelectPin, self)._set_state(value)
+        if self.spi_device:
+            self.spi_device.on_select()
+
+
+class MockSPIDevice(object):
+    def __init__(
+            self, clock_pin, mosi_pin, miso_pin, select_pin=None,
+            clock_polarity=False, clock_phase=False, lsb_first=False,
+            bits_per_word=8, select_high=False):
+        self.clock_pin = MockSPIClockPin(clock_pin)
+        self.mosi_pin = None if mosi_pin is None else MockPin(mosi_pin)
+        self.miso_pin = None if miso_pin is None else MockPin(miso_pin)
+        self.select_pin = None if select_pin is None else MockSPISelectPin(select_pin)
```

Comment:
Erm ... it should probably check clock_pin is not none, but I've been doing a bit of reading about SPI. Turns out there's some chips that do away with the MOSI or MISO pin entirely. For example, the MCP3xx1 chips don't bother with a MOSI pin at all (because they don't read any data, they simply return it). There's other chips that don't bother with a select line (they assume they're the only thing in the system). So the only guaranteed mandatory pin is clock (strictly clock and one of MOSI or MISO). Should probably implement some slightly more complex test here...